dgraphtest/cluster.go

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright 2023 Dgraph Labs, Inc. and Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dgraphtest
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	docker "github.com/docker/docker/client"
+	"github.com/pkg/errors"
+)
+
+var (
+	requestTimeout = 30 * time.Second
+	stopTimeout    = time.Minute
+)
+
+type cnet struct {
+	id   string
+	name string
+}
+
+type Cluster struct {
+	conf ClusterConfig
+
+	// resources
+	dcli   *docker.Client
+	net    cnet
+	zeros  []dnode
+	alphas []dnode
+}
+
+func NewCluster(conf ClusterConfig) (*Cluster, error) {
```

Comment:
If you return a pointer, I have seen it returned as a named param. But I don't think it has any operational difference.

Found this while reviewing -- https://medium.com/@philpearl/bad-go-pointer-returns-340f2da8289
(just instructive, do not have an issue with the signature)

Reply:
This is a good idea. We already have lots of pointers inside the struct. We don't need the Cluster to be a pointer then.